### DISCUSSION TITLE
Jetson Nano mender not updating dtb

### DISCUSSION THREAD

**Discussion by tirankas** [OE4T/meta-tegra#798] [General]
Hey Guys,
I now see a problem when updating the Jetson Nano using the mender.
Update process completes successfully, but dtb is not being updated.
I am using the dunfell branch.

nv_boot_control.conf shows this:
TNSPEC 3448-400-0002-F.0-1-0-jetson-nano-esd-imx-cot-mmcblk0p

Checking the bl_update_payload, I see this:
W>=3448-0000-100
>=3448-0002-100
W>=3448-0000-101
>=3448-0002-101
>=3448-0003-400
W<3448-0000-200
<3448-0002-200
W>=3448-0002-100
W<3448-0000-300
<3448-0002-300
>=3448-0003-000
W>=3448-0000-300
>=3448-0002-300
W>=3448-0003-400
W>=3448-0000-300
>=3448-0002-300
>=3448-0003-400
W<3448-0000-200

I don't see error with the u-boot-bup-payload deploy process though.

What am I missing?

Thanks

**Comment by ichergui** [comment]
Hey @tirankas 
It's hard to say anything. Are you using `Jetson Nano (eMMC version)` or `Jetson Nano (SDcard version)` ?
@madisongh  any thoughts ?

**Comment by tirankas** [comment]
Hi @ichergui ,
Using Jetson Nano (eMMC version).

Thanks

**Comment by madisongh** [comment]
Are you using the demo distro, and if so, what branch?   IIRC, meta-mender-tegra doesn't include bootloader updates for the t210 platforms (e.g., Nano).  We have some additions in the demo distro to add that, though.

**Comment by tirankas** [comment]
Hi @madisongh 
I am not using the demo distro directly but I did integrate most of it to our own layer.
I am using the dunfell 32.5.0 branch.
Can you please pinpoint to changes done in the demo distro to allow bootloader update? Maybe I can start by checking some lead there...

Thanks

**Comment by madisongh** [comment]
Take a look at the bbappend [here](https://github.com/OE4T/tegra-demo-distro/tree/dunfell-l4t-r32.5.0/layers/meta-tegra-support/dynamic-layers/meta-mender-tegra/recipes-mender/tegra-state-scripts).

**Comment by tirankas** [comment]
Hi @madisongh 
I actually have this change but I don't see this state script running during the mender update.
I should see mender runs this during update, shouldn't I?

Thanks

**Comment by madisongh** [comment]
Ah, oops.   I pointed you at the dunfell-l4t-r32.5.0 branch copy of that bbappend, but you're right, that still doesn't install the script for tegra210 platforms.  That got added to later branches, along with the switchover to using tegra-boot-tools instead of the NVIDIA update engine.   I think the important change is [here](https://github.com/OE4T/tegra-demo-distro/blob/4c27b8405b37a66e2deae4cf445fe5e7208d7a6e/layers/meta-tegra-support/dynamic-layers/meta-mender-tegra/recipes-mender/tegra-state-scripts/tegra-state-scripts_%25.bbappend#L18), if you want to try back-porting that to your branch.

**Comment by tirankas** [comment]
Thanks @madisongh, this makes sense now, will try this out...

**Comment by tirankas** [comment]
Hi @madisongh 
Well, tried this, had also to pull in the tegra-boot-tools-updater to get the tegra-bootloader-update tool.
Mender update completes successfully now but device is not powering up (no output from console too, probably means it doesn't find Uboot).
Was there anything else like uboot location change?

Is there a clean dunfell compatible source tree I can simply build on?

Thanks

**Comment by madisongh** [comment]
No output on the serial console, even?  It would help to know more about why it's not booting.

Was the device already running a build off your dunfell-l4t-r32.5.0 tree, or some other version? If you're trying to use a straight OTA upgrade between BSP versions, more care is needed to ensure compatibility, since the boot partition layout probably changed.

**Comment by tirankas** [comment]
Now I do see something on the console:

[0000.237] [L4T TegraBoot] (version 00.00.2018.01-l4t-dd84d362)
[0000.242] Processing in cold boot mode Bootloader 2
[0000.247] A02 Bootrom Patch rev = 1023
[0000.251] Power-up reason: software reset
[0000.254] No Battery Present
[0000.257] pmic max77620 reset reason
[0000.260] pmic max77620 NVERC : 0x0
[0000.264] RamCode = 0
[0000.266] Platform has DDR4 type RAM
[0000.269] max77620 disabling SD1 Remote Sense
[0000.274] Setting DDR voltage to 1125mv
[0000.278] Serial Number of Pmic Max77663: 0x329c6
[0000.285] Entering ramdump check
[0000.288] Get RamDumpCarveOut = 0x0
[0000.291] RamDumpCarveOut=0x0,  RamDumperFlag=0xe59ff3f8
[0000.296] Last reboot was clean, booting normally!
[0000.301] Sdram initialization is successful 
[0000.305] SecureOs Carveout Base=0x00000000ff800000 Size=0x00800000
[0000.311] Lp0 Carveout Base=0x00000000ff780000 Size=0x00001000
[0000.317] BpmpFw Carveout Base=0x00000000ff700000 Size=0x00080000
[0000.323] GSC1 Carveout Base=0x00000000ff600000 Size=0x00100000
[0000.329] GSC2 Carveout Base=0x00000000ff500000 Size=0x00100000
[0000.335] GSC4 Carveout Base=0x00000000ff400000 Size=0x00100000
[0000.340] GSC5 Carveout Base=0x00000000ff300000 Size=0x00100000
[0000.346] GSC3 Carveout Base=0x000000017f300000 Size=0x00d00000
[0000.362] RamDump Carveout Base=0x00000000ff280000 Size=0x00080000
[0000.369] Platform-DebugCarveout: 0
[0000.372] Nck Carveout Base=0x00000000ff080000 Size=0x00200000
[0000.378] Non secure mode, and RB not enabled.
[0000.394] Csd NumOfBlocks=0
[0000.632] *** Booting BFS0.
[0000.635] Read PT from (0:3)
[0000.641] Using BFS PT to query partitions 
[0000.645] PT: Partition LNX NOT found ! 
[0001.337] *** Failing over to KFS1.
[0000.286] [L4T TegraBoot] (version 00.00.2018.01-l4t-dd84d362)
[0000.292] Processing in cold boot mode Bootloader 2
[0000.296] A02 Bootrom Patch rev = 1023
[0000.300] Power-up reason: software reset
[0000.304] No Battery Present
[0000.306] pmic max77620 reset reason
[0000.310] pmic max77620 NVERC : 0x0
[0000.313] RamCode = 0
[0000.315] Platform has DDR4 type RAM
[0000.319] max77620 disabling SD1 Remote Sense
[0000.323] Setting DDR voltage to 1125mv
[0000.327] Serial Number of Pmic Max77663: 0x329c6
[0000.334] Entering ramdump check
[0000.337] Get RamDumpCarveOut = 0x0
[0000.341] RamDumpCarveOut=0x0,  RamDumperFlag=0xe59ff3f8
[0000.346] Last reboot was clean, booting normally!
[0000.350] Sdram initialization is successful 
[0000.354] SecureOs Carveout Base=0x00000000ff800000 Size=0x00800000
[0000.361] Lp0 Carveout Base=0x00000000ff780000 Size=0x00001000
[0000.366] BpmpFw Carveout Base=0x00000000ff700000 Size=0x00080000
[0000.372] GSC1 Carveout Base=0x00000000ff600000 Size=0x00100000
[0000.378] GSC2 Carveout Base=0x00000000ff500000 Size=0x00100000
[0000.384] GSC4 Carveout Base=0x00000000ff400000 Size=0x00100000
[0000.390] GSC5 Carveout Base=0x00000000ff300000 Size=0x00100000
[0000.395] GSC3 Carveout Base=0x000000017f300000 Size=0x00d00000
[0000.412] RamDump Carveout Base=0x00000000ff280000 Size=0x00080000
[0000.418] Platform-DebugCarveout: 0
[0000.421] Nck Carveout Base=0x00000000ff080000 Size=0x00200000
[0000.427] Non secure mode, and RB not enabled.
[0000.443] Csd NumOfBlocks=0
[0000.675] *** Booting BFS1.
[0000.677] Read PT from (0:3)
[0000.683] Using BFS PT to query partitions 
[0000.687] PT: Partition LNX NOT found ! 
[0001.376] *** Failing over to KFS2.

**Comment by tirankas** [comment]
Also, before the update running this:
fw_printenv -n mender_boot_part
Configuration file wrong or corrupted

**Comment by madisongh** [comment]
Again:

> Was the device already running a build off your dunfell-l4t-r32.5.0 tree, or some other version?

From the messages you mention, it sounds like there is a mismatch between the existing partition layout and what you're trying to update to.  And likely a mismatch in the u-boot builds, with different locations for the environment block.

**Reply by tirankas** [reply]
Was the device already running a build off your dunfell-l4t-r32.5.0 tree, or some other version? --> Yes, already running 32.5.0 dunfell branch

**Reply by madisongh** [reply]
Hmm. Have you made any other changes besides adding the state script?

**Comment by tirankas** [comment]
Hi @madisongh 
I picked up a bunch of commits you made from [here](https://github.com/OE4T/tegra-demo-distro/pull/32) 

The device is currently running with an image before this change and the mender I am trying to apy is with these commits.

Does above commits causes uboot change in a way that it cannot be used with an image that doesn't have these changes?

**Reply by madisongh** [reply]
IIRC, that initial set of commits covered the cboot platforms.  There were additional changes needed for t210/u-boot platforms (commit hashes d460cf6, a6254c8) you might need.

**Comment by tirankas** [comment]
Thanks @madisongh, will verify these fixes...
Btw, aren't all these changes upstream? I don't see these changes in the meta tegra in the master as well.

Thanks

**Reply by madisongh** [reply]
They would go into the mender-mender-community repository under meta-mender-tegra, and I'm not sure what the migration plan is there for dealing with updates beyond L4T R32.4.3. @dwalkes ?

**Reply by dwalkes** [reply]
@madisongh good question I don't think we've specifically discussed this.  However given we are [using JP 4.6 for the dunfell branch of tegra-demo-distro](https://github.com/OE4T/meta-tegra/wiki/OE4T-Meeting-Notes-2021-08-12) I would expect we probably want to upstream the same fixes to meta-mender-community.  For now, managing in our dynamic layers has worked and I don't expect people who use mender + yocto on Tegra and want community support would expect to deviate significantly from what we do in tegra-demo-distro.

However, regarding https://github.com/OE4T/tegra-demo-distro/pull/32 isn't the question essentially whether we want to upstream not only support for R32.6 but also if we want to upstream a switch from NVIDIA bootloader tools to tegra-boot-tools on the dunfell branch?

If I look at https://docs.google.com/spreadsheets/d/1EWG5QZvtWdk74gz48D4xCbegK-bsqMFJ8XOsyGdUUeI/edit#gid=145133208 we have coverage of mender update testing on tegra-demo-distro jetson-nano-devkit-emmc on R32.6.1 (row 18) - @ichergui was this upgrade going to the master branch or the dunfell branch of tesgra demo distro?  Has anyone specifically tested tegra-demo-distro upgrades from 32.5 to `dunfell` (32.6) branch on tegra-demo-distro?

@tirankas can you reproduce this problem with tegra demo distro and detail the combination which produces a failure in a test case which looks like one of the ones in https://docs.google.com/spreadsheets/d/1EWG5QZvtWdk74gz48D4xCbegK-bsqMFJ8XOsyGdUUeI/edit#gid=145133208?  I think this data could help us decide how to resolve.

Is the issue here related to the comments in [the July meeting](https://github.com/OE4T/meta-tegra/wiki/OE4T-Meeting-Notes-2021-07-08) about whether we should have changed uboot environment in nano?  I don't think I fully understand the issue from the comments here, but it looks like it's something related to either moving uboot environment or changing partition locations or both.

My inclination is not to attempt to switch the dunfell branch to use tegra-boot-tools unless there's a specific reason we need to.  In the [July meeting](https://github.com/OE4T/meta-tegra/wiki/OE4T-Meeting-Notes-2021-07-08) we mentioned that for cross L4T upgrade problems on Nano our first/preferred approach was going to be to update documentation, warning/error messages.

@atharvanan1 you are more in tune with the state of the branches on tegra-demo-distro than I am at this point, do you have thoughts about the correct path here regarding NVIDIA bootloader tools switchover on the dunfell tegra-demo-distro branch?

**Reply by tirankas** [reply]
Hi @dwalkes 
My setup is based on the dunfell 32.5.0 branch of the Tegra demo.
One of the changes I did was to include a support for a new camera we added which causes a change in the DT.
Flashing my build with the updated DTB, I see DTB is not applied with mender which leads to my discussion with Matt. There is no error though, just DTB stays as before...
Are you saying the 32.5.0 dunfell branch of the Tegra demo distro should handle DTB update via mender?
I can test from scratch if you tell me it does.

Thanks

**Reply by dwalkes** [reply]
OK that makes sense.

> Are you saying the 32.5.0 dunfell branch of the Tegra demo distro should handle DTB update via mender?

No, I suspect it probably doesn't and that's the reason @madisongh recommended https://github.com/OE4T/tegra-demo-distro/pull/32

So the test case to reproduce with tegra-demo-distro would be:
1. Start with the r32.5.0 dunfell branch.
3. Change the DTB.
3. Perform the upgrade, the DTB is not applied.

Is that correct?

If so I'd suggest documentation about how to handle this rather than changing the default bootloader on the dunfell branch of tegra-demo-distro or elsewhere.  And I'd suggest that documentation maybe references a configuration or set of patches which would switch the boot tools to use `tegra-boot-tools` on dunfell.

I could be convinced otherwise by people who are more knowledgeable about this than I am, which would include at least @madisongh @ichergui, and @atharvanan1 and probably others.  The main concern I'd have is someone who pulls the latest `dunfell` branch of `tegra-demo-distro` (but didn't change their DTB) and suddenly has a new set of bootloader tools on their next upgrade.  I know I haven't tested enough of the bootloader tools switch myself to be confident this wouldn't introduce problems.  If we have a new Kirkstone LTS branch early next summer this may be a time to introduce the switch to new bootloader tools, for long term support on the end of the road JP 4.6 release for Nano and TX2

**Reply by tirankas** [reply]
Hi @dwalkes 
The steps above are correct, this is exactly what I did.
I will do this from clean, to make sure I am fetch the up-to-date repo and will report back here.
I understand your concern, but am not sure which path is better: stay with an issue like this (maybe people think boot was updated while it doesn't and will find this maybe too late), or fixing it to make boot change and align to what should be. Tough question....

Thanks

**Reply by dwalkes** [reply]
> maybe people think boot was updated while it doesn't and will find this maybe too late

True, but I'm guessing in most cases if they do find it late, hopefully they could still find/use documentation which tells them how to resolve it using another update and it won't brick their device.  Hopefully any brick type of scenarios would get caught by their testing, as you've caught in your case.

**Reply by atharvanan1** [reply]
> @atharvanan1 you are more in tune with the state of the branches on tegra-demo-distro than I am at this point, do you have thoughts about the correct path here regarding NVIDIA bootloader tools switchover on the dunfell tegra-demo-distro branch?

Here's what the boot-tools look like in different tegra-demo-distro branches:
| Branch | Tools |
|----------|--------|
| dunfell-l4t-r32.4.3 | tegra-boot-tools |
| dunfell-l4t-r32.5.0 | tegra-redundant-boot |
| dunfell | tegra-redundant-tools |
| master | tegra-boot-tools |

If anyone is trying to move from latest tegra-demo-distro `dunfell-l4t-r32.4.3` to `dunfell` (32.6), they have already tested that tegra-boot-tools and nv_update_engine work together. @dwalkes, we never tested this as we don't have the `tegra-boot-tools` in our internal company branches as we haven't pulled latest contents from `tegra-demo-distro` after Feb 25. If anyone tested through an upgrade path from tegra-redundant-boot to tegra-boot-tools and vice versa, I'm not sure if they ever found any issues jumping between both. This is key to understanding bootloader strategy we would want to use with these braches. If jumping works fine, I'd say we should change to `tegra-boot-tools`.

Our trouble with `jetson-nano-emmc` was that VER partition was overwritten by U-boot environment. This caused `tegra-bootloader-update` to error out, we added a fix for giving us a warning after  we update on that system. I probably need to look into what's the partition location across different branches, but pre-dunfell-l4t-r32.4.2 and dunfell-l4t-r32.4.3 there has been a change.

**Reply by dwalkes** [reply]
Thanks for clarifying @atharvanan1 .  It seems confusing to use `tegra-boot-tools` on r32.4.3 but `tegra-redundant-boot` on dunfell.

This means if someone was using `tegra-demo-distro` branch `dunfell-l4t-r32.4.3` previously they will go through a transition of 
bootloader tools as well, correct?  It also means someone going from `dunfell-l4t-r32.4.3` to `dunfell` (32.6) they would have a transition in the wrong direction from NVIDIA bootloader tools to the `tegra-boot-tools` we'd like to use going forward.  So I think it probably makes more sense to use `tegra-boot-tools` on `dunfell`

**Reply by madisongh** [reply]
So we're using tegra-boot-tools in dunfell-l4t-r32.4.3  (of the demo distro) because @atharvanan1 graciously back-ported that change. That change never got re-ported to dunfell-l4t-r32.5.0, and the R32.6.1 support in dunfell was based on the R32.5.0 branch.  That's the missing link, I guess.... one of the disadvantages of juggling so many parallel branches.

**Reply by dwalkes** [reply]
Got it.  So @tirankas if you are willing to support a port to either/both of the dunfell and dunfell-l4t-r32.5.0 this looks like what we need.  @atharvanan1 can help you as needed.  When this is complete we should update mender hub pages and the upstream meta-mender-community with the content from the dunfell-l4t-r32.4.3 branch and reference the long term support strategy on the dunfell branch on tegra-demo-distro to use r32.6.  When Nvidia releases deepstream for 32.6 we might want to update meta-mender-community branch to use dunfell 32.6 instead.  Does this seem like the right approach?

**Reply by madisongh** [reply]
Sounds good to me.

**Reply by tirankas** [reply]
Hi @dwalkes,
Sure, will help as needed.
Hi @madisongh,
Let me know what you need from me and I will follow up.
I have made the required changes based on dunfell 32.5.0 and verified it's working fine. However, I made these changes in my own layer, not touching the tegra demo. If you need, I can re-arrange those changes and submit to tegrademo and whatever layer required.
Please let me know what help you need from my side.

Thanks

**Reply by dwalkes** [reply]
>  If you need, I can re-arrange those changes and submit to tegrademo and whatever layer required.

It may make sense to submit to `dunfell-l4t-r32.5.0` branch on tegra-demo-distro and add @atharvanan1 as a reviewer, if that's the one which is ready now.  Then we could port from there into `dunfell` and the r32.6 release.

**Reply by tirankas** [reply]
Sure @dwalkes, will do.
I am rebasing my changes directly on the tegrademo (on the 32.5.0 branch). I will then compile and submit.
Thanks

**Reply by tirankas** [reply]
Hi @madisongh
I have the commit ready, tested on the Jetson Nano (Emmc based) and it works fine.
However, during testing on the NX-Xavier I see the devices crashes (update went fine, reboot after update crashes and device reverts back), this is coming from Kernel/Init, I don't have the logs in hand (will only have them when back to work on Wednesday).
Do you want me to send the commit for inspection now or shall I wait till figure out the problem?

Thanks

**Reply by madisongh** [reply]
Go ahead an submit the PR.  With more eyes on it, we should be able to resolve the issue more quickly.  Thanks!

**Reply by tirankas** [reply]
Thanks @madisongh 
Can someone give me write access to the tegrademo repo? I get permission denied trying to push

**Reply by atharvanan1** [reply]
Hey @tirankas, without the write access, you can fork `tegra-demo-distro`. You won't need write access in that case. Creating a new branch and generating a PR can be done against the "original" branch through the fork too.

**Reply by madisongh** [reply]
The workflow would be to fork the repo to one under your own account, push your changes there, then submit the PR.

**Reply by tirankas** [reply]
Thanks Guys. Done: PR [here](https://github.com/OE4T/tegra-demo-distro/pull/154/commits/6a582f6df6866ca1d27cc51217bdc33d51862ec9)

**Reply by madisongh** [reply]
The crashing problem on the Xavier is due to a missing newline at the end of the `platform-preboot-cboot.sh` script installed by the `tegra-minimal-init` bbappend.  I see a handful of other gratuitous whitespace changes and missing end-of-file newlines across the change set.  Easy enough to fix up.

**Reply by tirankas** [reply]
Hi @madisongh 
With the latest release, when I update Jetson Nano (EMMC) from pervious version to this new version via the mender server, I get this error:
2021-09-21 14:04:08 +0000 UTC info: Executing script: ArtifactInstall_Leave_80_bl-update 
2021-09-21 14:04:09 +0000 UTC info: Collected output (stderr) while running script /var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update /dev/disk/by-partlabel/BMP: No space left on device ERR: bootloader update failed  
---------- end of script output 
2021-09-21 14:04:09 +0000 UTC error: transient error: error executing leave script for update-install state: error running leave state script(s) for ArtifactInstall state: statescript: error executing 'ArtifactInstall_Leave_80_bl-update': 1 : exit status 1 
2021-09-21 14:04:09 +0000 UTC info: State transition: update-install [ArtifactInstall] -> rollback [ArtifactRollback]

**Reply by madisongh** [reply]
Compare the flash layout XML files between whatever your previous release contained vs. the one for whatever you are calling latest.  When I look at the flash layouts for the tegrademo-mender distro for dunfell-l4t-r32.4.3 vs. dunfell-l4t-r32.5.0 vs. dunfell (R32.6.1), the sizes are identical across the board.  If you're using your own distro, perhaps the BMP partition is smaller in the later build vs. the earlier one (or vice-versa)?

**Reply by tirankas** [reply]
Hi @madisongh 
I checked and they are identical.
Also, second time I try to update I get a different thing:
2021-09-21 14:23:45 +0000 UTC info: Executing script: ArtifactInstall_Leave_80_bl-update 
2021-09-21 14:23:45 +0000 UTC info: Collected output (stderr) while running script /var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update ERR: cannot perform bootloader update  
---------- end of script output 
2021-09-21 14:23:45 +0000 UTC error: transient error: error executing leave script for update-install state: error running leave state script(s) for ArtifactInstall state: statescript: error executing 'ArtifactInstall_Leave_80_bl-update': 1 : exit status 1

**Reply by madisongh** [reply]
From which branch is the build that is currently running, and from which is the one you're trying to upgrade to?  I can try and run some tests with the demo distro to make sure nothing got lost during the merges, but I did verify that updates for the Nano-eMMC worked across the three dunfell variants we have there.

If you can manually run `bash -x /var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update`, that would be helpful.

**Reply by tirankas** [reply]
Both versions are on the dunfell 32.5.0 branch.
I willsend the output soon.

**Reply by tirankas** [reply]
Hi @madisongh 
You want me to run this on the existing version (without actually updating)?

**Reply by madisongh** [reply]
Run it after the update fails.  The script (which got downloaded with the Mender artifact) should still be there.

**Reply by tirankas** [reply]
Hi @madisongh 
Here is the output:

+ UBOOTVARS='bootcount mender_boot_part mender_boot_part_hex upgrade_available'
+ ubvarfile=
+ install_fwenv=
+ echo 'Installing NVIDIA bootloader update payload'
Installing NVIDIA bootloader update payload
++ fw_printenv -n mender_boot_part
+ new_boot_part=1
++ mktemp -d -t nvbup.XXXXXX
+ mnt=/tmp/nvbup.kHll9y
+ '[' -z /tmp/nvbup.kHll9y -o '!' -d /tmp/nvbup.kHll9y ']'
+ mount -o ro /dev/mmcblk0p1 /tmp/nvbup.kHll9y
+ '[' '!' -d /tmp/nvbup.kHll9y/opt/ota_package ']'
+ mount --bind /sys /tmp/nvbup.kHll9y/sys
+ mount --bind /proc /tmp/nvbup.kHll9y/proc
+ mount --bind /dev /tmp/nvbup.kHll9y/dev
+ mount -t tmpfs tmpfs /tmp/nvbup.kHll9y/run
+ '[' -e /tmp/nvbup.kHll9y/usr/share/u-boot/fw_env.config ']'
+ chroot /tmp/nvbup.kHll9y /usr/bin/tegra-bootloader-update --dry-run /opt/ota_package/bl_update_payload
+ chroot /tmp/nvbup.kHll9y /usr/bin/tegra-bootloader-update --dry-run /opt/ota_package/bl_update_payload
+ grep -q 'version partitions are corrupted'
+ echo 'ERR: cannot perform bootloader update'
ERR: cannot perform bootloader update
+ cleanup
+ '[' -n /tmp/nvbup.kHll9y ']'
+ for d in sys proc dev run
+ mountpoint -q /tmp/nvbup.kHll9y/sys
+ umount /tmp/nvbup.kHll9y/sys
+ for d in sys proc dev run
+ mountpoint -q /tmp/nvbup.kHll9y/proc
+ umount /tmp/nvbup.kHll9y/proc
+ for d in sys proc dev run
+ mountpoint -q /tmp/nvbup.kHll9y/dev
+ umount /tmp/nvbup.kHll9y/dev
+ for d in sys proc dev run
+ mountpoint -q /tmp/nvbup.kHll9y/run
+ umount /tmp/nvbup.kHll9y/run
+ mountpoint -q /tmp/nvbup.kHll9y
+ umount /tmp/nvbup.kHll9y
+ rmdir /tmp/nvbup.kHll9y
+ exit 1

**Reply by tirankas** [reply]
Please ignore my last post, it was with the old script, not new

**Reply by madisongh** [reply]
BTW, if this is your own distro - did you update meta-tegra to the latest on dunfell-l4t-r32.5.0 also?  I did have to update tegra-boot-tools to fix an issue with the Nano updates, so if you don't have version 2.5.1 of that package, be sure to update meta-tegra to latest on that branch to pick up that change.

**Reply by tirankas** [reply]
I probably didn't, picked up the repos as it is via the tegrademo repo. Is this sufficient to only pick up the 2.5.1 of the tegra boot tools package? Less changes, less risk.

**Reply by tirankas** [reply]
I checked and I do have 2.5.1 of the boot tools.

**Reply by tirankas** [reply]
Hi @madisongh 
Here is the output of that script:
 sudo bash -x /var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update 
+ mnt=
+ UBOOTVARS='bootcount mender_boot_part mender_boot_part_hex upgrade_available'
+ ubvarfile=
+ install_fwenv=
+ echo 'Installing NVIDIA bootloader update payload'
Installing NVIDIA bootloader update payload
++ fw_printenv -n mender_boot_part
+ new_boot_part=1
++ mktemp -d -t nvbup.XXXXXX
+ mnt=/tmp/nvbup.3LXvXI
+ '[' -z /tmp/nvbup.3LXvXI -o '!' -d /tmp/nvbup.3LXvXI ']'
+ mount -o ro /dev/mmcblk0p1 /tmp/nvbup.3LXvXI
+ '[' '!' -d /tmp/nvbup.3LXvXI/opt/ota_package ']'
+ mount --bind /sys /tmp/nvbup.3LXvXI/sys
+ mount --bind /proc /tmp/nvbup.3LXvXI/proc
+ mount --bind /dev /tmp/nvbup.3LXvXI/dev
+ mount -t tmpfs tmpfs /tmp/nvbup.3LXvXI/run
+ '[' -e /tmp/nvbup.3LXvXI/usr/share/u-boot/fw_env.config ']'
+ chroot /tmp/nvbup.3LXvXI /usr/bin/tegra-bootloader-update --dry-run /opt/ota_package/bl_update_payload
+ chroot /tmp/nvbup.3LXvXI /usr/bin/tegra-bootloader-update --dry-run /opt/ota_package/bl_update_payload
+ grep -q 'version partitions are corrupted'
+ echo 'ERR: cannot perform bootloader update'
ERR: cannot perform bootloader update
+ cleanup
+ '[' -n /tmp/nvbup.3LXvXI ']'
+ for d in sys proc dev run
+ mountpoint -q /tmp/nvbup.3LXvXI/sys
+ umount /tmp/nvbup.3LXvXI/sys
+ for d in sys proc dev run
+ mountpoint -q /tmp/nvbup.3LXvXI/proc
+ umount /tmp/nvbup.3LXvXI/proc
+ for d in sys proc dev run
+ mountpoint -q /tmp/nvbup.3LXvXI/dev
+ umount /tmp/nvbup.3LXvXI/dev
+ for d in sys proc dev run
+ mountpoint -q /tmp/nvbup.3LXvXI/run
+ umount /tmp/nvbup.3LXvXI/run
+ mountpoint -q /tmp/nvbup.3LXvXI
+ umount /tmp/nvbup.3LXvXI
+ rmdir /tmp/nvbup.3LXvXI
+ exit 1

Thanks

**Reply by madisongh** [reply]
I've finally got a chance to go back and re-verify Mender updates on the Nano-eMMC platform with the demo distro, and all updates are still working OK across the three dunfell branches.

Looking at the output you posted, I'd say that the updater detected that the version partitions are corrupted, so there is, in fact, a difference in the boot partition layout between the currently running version and the new version.  The script will attempt to work around that if it finds a copy of the U-Boot `fw_env.config` file in `/usr/share/u-boot` on the new rootfs (performed [here](https://github.com/OE4T/tegra-demo-distro/blob/f3d9f9a5f89abcec44ff07477a1e98e1cebcefa1/layers/meta-tegra-support/dynamic-layers/meta-mender-tegra/recipes-bsp/u-boot/libubootenv_%25.bbappend#L21) in the libubootenv bbappend).  From the log output, it looks like that file is *not* installed in the new rootfs.  Can you check on that?

**Comment by tirankas** [comment]
Thanks @madisongh 
What is their strategy here?
If I take the master branch from the community tegra part, what should I merge in to have the latest and greatest? Would I be facing compatibility issues with other layers if I go that path?

**Reply by dwalkes** [reply]
> If I take the master branch from the community tegra part,

Do you mean the master branch of `tegra-demo-distro`?  

> Would I be facing compatibility issues with other layers if I go that path?

I think in general for mender L4T upgrades across releases if you don't see the path outlined in https://docs.google.com/spreadsheets/d/1EWG5QZvtWdk74gz48D4xCbegK-bsqMFJ8XOsyGdUUeI/edit#gid=145133208 you are blazing a new trail and your help to document/resolve issues is much appreciated.

**Reply by tirankas** [reply]
Hi @dwalkes 
By the wayz looking at the discussion for the jetpack 4.6 you guys had, I see one of the items is to move to squashfs.
One of the first changes I did when starting with the Jetsons was to move to squashfs. Aside from flash reduction we gained from this (which is significant), squashfs is proven to be very stable in terms of flash robustness. I have been using squashfs with many Qualcomm chipsets (mdm, snapdragons etc) as a FS for the automotive mainly due to the automotive strict flash robustness needs (15 years in field...). Maybe even adding ubi on top of it would be best as another layer of flexibility, ease of maintenance etc...

Btw, you will need to add squashfs support to uboot to get this to work...

**Reply by dwalkes** [reply]
Thanks @tirankas I've added a cross-link here from https://github.com/OE4T/meta-tegra/wiki/OE4T-Meeting-Notes-2021-08-12

**Comment by dwalkes** [comment]
Thanks @madisongh and @tirankas for all your work on https://github.com/OE4T/tegra-demo-distro/pull/154

It looks like there's one remaining issue in the thread here at https://github.com/OE4T/meta-tegra/discussions/798#discussioncomment-1374635 which is likely specific to @tirankas custom distro if I understand correctly.

It looks like tegra-demo-distro now uses `tegra-boot-tools` on all branches.  I started a wiki page at https://github.com/OE4T/tegra-demo-distro/wiki/Bootloader-tools-use-in-tegra-demo-distro  which links back here and references a bootloader change in existing branches, especially r32.4.3, to help highlight to people that they'll be switching bootloader tools if they rebase their branch of tegra-demo-distro, gives an overview of some of the motivation for this.  @atharvanan1 can you review when you have a chance and make sure I got this right?

**Reply by atharvanan1** [reply]
Yeah, I have verified that we are using tegra-boot-tools in all the branches now - in terms of using `tegra-boot-tools` in mender tools (can be seen in [here](https://github.com/OE4T/meta-tegra-community/blob/master/dynamic-layers/meta-mender-tegra/recipes-mender/tegra-state-scripts/files/redundant-boot-install-script)). I don't see [this change](https://github.com/OE4T/tegra-demo-distro/blob/dunfell-l4t-r32.4.3/layers/meta-tegra-support/recipes-bsp/tegra-bup-payload/tegra-bup-payload_%25.bbappend) anywhere I'm looking, but I might not be looking at correct place anymore.

**Reply by madisongh** [reply]
Yeah, that change (to package the full payload) is just in the r32.4.3 branch of the demo distro.

Rather than porting that around to the other branches, it might make sense to update the base recipe in meta-tegra to create a separate package containing that payload, then use that package when creating an image for use with the OTA re-flashing sequence.
